sycl/include/CL/sycl/detail/cstr.hpp

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+///  \file
+///  This file contains meta programming utilities to manipulate string
+///  constants Stored in a type like cstr<char, 'a', 'b'>. This is used for
+///  kernel properties because they must be constants and the API reflects that.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SYCL_DETAIL_CSTR_HPP
+#define SYCL_DETAIL_CSTR_HPP
+
+#include "CL/sycl/detail/pi.h"
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+
+namespace sycl {
+
+namespace detail {
+
+/// utility to make any functor's operator() be const even if it originally
+/// wasn't. This is useful to bypass the restriction that kernel operator() must
+/// be const.
+template <typename T> struct const_cheating_wrapper {
+  mutable std::decay_t<T> t;
+  template <typename... Ts> auto operator()(Ts &&... ts) const & {
+    return std::forward<T &>(t)(std::forward<Ts>(ts)...);
+  }
+  template <typename... Ts> auto operator()(Ts &&... ts) const && {
+    return std::forward<T &&>(t)(std::forward<Ts>(ts)...);
+  }
+};
+template <typename T> auto make_const_cheater(T &&t) {
+  return const_cheating_wrapper<T>{std::forward<T>(t)};
+}
+
+/// transform a type like: cstr<char, 'a', 'b'>
+/// into a string "ab"
+template <typename> struct StrPacker {};
+
+/// utility type to manipulate string in constant context
+template <typename CharT, CharT... charpack> struct cstr {
+  using Char = CharT;
+  static constexpr unsigned size = sizeof...(charpack);
+  template <unsigned Idx>
+  using at = __type_pack_element<Idx, cstr<CharT, charpack>...>;
+};
+
+template <typename CharT, CharT... P1> struct StrPacker<cstr<CharT, P1...>> {
+  static constexpr const CharT Str[] = {P1...};
+};
+
+/// utility to merge string with a separator
+template <typename...> struct concat {};
+
+template <typename CharT, CharT... Sep, CharT... P1>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>> {
+  using type = cstr<CharT, P1...>;
+};
+
+template <typename CharT, CharT... Sep, CharT... P1, CharT... P2,
+          typename... Rest>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>, cstr<CharT, P2...>,
+              Rest...> {
+  using type =
+      typename concat<cstr<CharT, Sep...>, cstr<CharT, P1..., Sep..., P2...>,
+                      Rest...>::type;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+/// We are using a compiler extention here and we don't want any warning about
+/// it so we suppress that warning.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-string-literal-operator-template"
+
+/// utility to transform a string literal into its cstr form.
+template <typename CharT, CharT... charpack>
+__SYCL_ALWAYS_INLINE constexpr cstr<CharT, charpack...>
+operator"" _cstr() noexcept {
+  return {};
+}
+#pragma clang diagnostic pop
+
+using Base2 = decltype("01"_cstr);
```

Comment:
Perhaps you can coin some design comments here. Like this has 3 roles:
- as a tag to select a base;
- as a description of each character to use for each "digit";
- as the base itself by just counting the number of digit characters.

Do not forget the readers are not telepathic or you. :-)